src/serac/infrastructure/accelerator.hpp

```diff
@@ -247,6 +253,29 @@ std::shared_ptr<T[]> make_shared_array(std::size_t n)
 #endif
 }
 
+/**
+ * @brief create shared_ptr to an array of `n` values of type `T`, either on the host or device
+ * @tparam T the type of the value to be stored in the array
+ * @tparam exec the memory space where the data lives
+ * @param n how many entries to allocate in the array
+ */
+template <ExecutionSpace exec, typename... T>
+auto make_shared_arrays(std::size_t n)
+{
+  if constexpr (exec == ExecutionSpace::CPU) {
+    return std::tuple{std::shared_ptr<T[]>(new T[n])...};
+  }
+
+#if defined(__CUDACC__)
+  if constexpr (exec == ExecutionSpace::GPU) {
+    T* data;
+    cudaMalloc(&data, sizeof(T) * n);
+    auto deleter = [](T* ptr) { cudaFree(ptr); };
+    return std::shared_ptr<T[]>(data, deleter);
+  }
+#endif
```

Comment:
I believe this will never happen with the current implementation, since the GPU specialization of the code that calls this function is gated behind some conditional compilation to turn it off when serac is not configured with cuda support.

However, we could do something like

```cpp
/**
 * @brief enum used for signalling whether or not to perform certain calculations on the CPU or GPU
 */
enum class ExecutionSpace
{
  CPU,
#ifdef SERAC_CUDA_ENABLED
  GPU,
  Dynamic  // Corresponds to execution that can "legally" happen on either the host or device
#endif
};
```

To explicitly fail everywhere if a user tries to ask for a GPU execution space without cuda support.

Reply:
I think that's a good idea just as a failsafe. Any thoughts @white238 @adayton1 ?

Reply:
I think adding another define would be confusing to figure out how it differs from the ones we have already.  This should be fine with just this:

```
#if defined(AXOM_USE_CUDA) && defined(__CUDACC__)
```